lib/nerve.rb

```diff
@@ -15,35 +15,103 @@ class Nerve
 
     include Logging
 
-    def initialize(opts={})
+    def initialize(config_manager)
       log.info 'nerve: starting up!'
+      @config_manager = config_manager
 
       # set global variable for exit signal
       $EXIT = false
 
+      @watchers = {}
+      @watcher_versions = {}
+
+      # Flag to indicate a config reload is required by the main loop
+      # This decoupling is required for gracefully reloading config on SIGHUP
+      # as one should do basically nothing in a signal handler
+      @config_to_load = true
+
+      # Will be populated by load_config! in the main loop
+      @instance_id = nil
+      @watchers_desired = {}
+      @heartbeat_path = nil
+
+      Signal.trap("HUP") do
+        @config_to_load = true
+      end
+
+     log.debug 'nerve: completed init'
+    end
+
+    def load_config!
+      log.info 'nerve: loading config'
+      @config_to_load = false
+      @config_manager.reload!
+      config = @config_manager.config
+
       # required options
       log.debug 'nerve: checking for required inputs'
       %w{instance_id services}.each do |required|
-        raise ArgumentError, "you need to specify required argument #{required}" unless opts[required]
+        raise ArgumentError, "you need to specify required argument #{required}" unless config[required]
       end
-
-      @instance_id = opts['instance_id']
-      @services = opts['services']
-      @heartbeat_path = opts['heartbeat_path']
-      @watchers = {}
-
-      log.debug 'nerve: completed init'
+      @instance_id = config['instance_id']
+      @watchers_desired = config['services']
+      @heartbeat_path = config['heartbeat_path']
     end
 
     def run
       log.info 'nerve: starting run'
-
-      @services.each do |name, config|
-        launch_watcher(name, config)
-      end
-
       begin
         loop do
+          # Check if configuration needs to be reloaded and reconcile any new
+          # configuration of watchers with old configuration
+          if @config_to_load
+            load_config!
+
+            # Reap undesired service watchers
+            services_to_reap = @watchers.select{ |name, _|
+              !@watchers_desired.has_key?(name)
+            }.keys()
+            log.info "nerve: reaping old watchers: #{services_to_reap}"
```

Comment:
could probably avoid outputting that log line unless we actually are going to reap something.

```ruby
unless services_to_reap.empty?
  log.info "..."
  services_to_reap.each {}
end
```